Stripe/STPPostalCodeValidator.m

```diff
@@ -7,40 +7,171 @@
+            // All numeric postal code entered
+            if (numberOfDigits == 9) {
+                return STPCardValidationStateValid;
+            }
+            else {
+                return STPCardValidationStateIncomplete;
+            }
+
+        }
+        else if ((numberOfDigits + 1) == totalLength) {
+            // Possibly has a separator character for ZIP+4, check to see if
+            // its in the right place
+
+            NSString *separatorCharacter = [postalCode substringWithRange:NSMakeRange(5, 1)];
+            if (countOfCharactersFromSetInString(separatorCharacter, [NSCharacterSet stp_asciiDigitCharacterSet]) == 0) {
+                // Non-digit is in right position to be separator
+                if (numberOfDigits == 9) {
+                    return STPCardValidationStateValid;
+                }
+                else {
+                    return STPCardValidationStateIncomplete;
+                }
+            }
+            else {
+                // Non-digit is in wrong position to be separator
+                return STPCardValidationStateInvalid;
+            }
+        }
+        else {
+            // Not a valid zip code (too many non-numeric characters)
+            return STPCardValidationStateInvalid;
+        }
+    }
+}
++ (NSString *)formattedSanitizedPostalCodeFromString:(NSString *)postalCode
+                                         countryCode:(NSString *)countryCode
+                                               usage:(STPPostalCodeIntendedUsage)usage {
+    if (countryCode == nil) {
+        return postalCode;
+    }
+
+    NSString *sanitizedCountryCode = countryCode.uppercaseString;
+    if ([sanitizedCountryCode isEqualToString:STPCountryCodeUnitedStates]) {
+        return [self formattedSanitizedUSZipCodeFromString:postalCode
+                                                     usage:usage];
+    }
+    else {
+        return postalCode;
+    }
+
+}
+
++ (NSString *)formattedSanitizedUSZipCodeFromString:(NSString *)zipCode
+                                              usage:(__unused STPPostalCodeIntendedUsage)usage {
+    NSUInteger maxLength = 0;
+    switch (usage) {
+        case STPPostalCodeIntendedUsageBillingAddress:
+            maxLength = 5;
+            break;
+        case STPPostalCodeIntendedUsageShippingAddress:
+            maxLength = 9;
+    }
+
+
+    NSString *formattedString = [[STPCardValidator sanitizedNumericStringForString:zipCode] stp_safeSubstringToIndex:maxLength];
+
+    if (formattedString.length > 5
+        || (formattedString.length == 5
+            && [[zipCode substringFromIndex:(zipCode.length - 1)] isEqualToString:@"-"])) {
+        NSMutableString *mutableZip = formattedString.mutableCopy;
+        [mutableZip insertString:@"-" atIndex:5];
+        formattedString = mutableZip.copy;
+    }
+
+    return formattedString;
+}
+
+
++ (BOOL)postalCodeIsRequiredForCountryCode:(NSString *)countryCode {
+    if (countryCode == nil) {
```

Comment:
Nonnil codes of any length will get caught in the else branch here and so don't necessarily need special logic like this. While i use use `.length == 0` for string checking, I thought it more clear to call it out like this since it's specifically checking that nil values can't be passed to `containsObject:`